### PR TITLE
some improve...

### PR DESCRIPTION
**Pull request by tianyiw2013**
Including but not limited to the following modifications:

- localization
  - package.nls.json `My English is poor, please check and improve the content`
  - package.nls.zh-cn.json `Simplified Chinese`
- configuration
  - php-docblocker.unknownType #177
    - `[type]`
    - 'mixed'
  - php-docblocker.withParameterDescription #164
    - true `@param ${2:string} \\$name ${3:name}`
    - false `@param ${2:string} \\$name ${3}`
  - php-docblocker.defaultMessage #139
    - undocumented `/** Undocumented class */`
    - blank
    - name `/** Test */ class Test`
  - php-docblocker.constTemplate
  - php-docblocker.enumTemplate
  - php-docblocker.variableTemplate
  - php-docblocker.variableInline #42
    - true
        `/** @var ${1:string} ${2:\\$var} ${3} */`
    - false
        ```php
        /** 
         * ${1:Undocumented variable}
         *
         * @var ${2:string} ${3:\\$var}
         **/
        ```
  - php-docblocker.variableWithKey
    - true
        `/** @var ${1:string} ${2:\\$var} ${3} */`
    - false
        `/** @var ${1:string} ${2} */`
- block
  - variable `$var = 1`
    - foreach
    ```php
      /** @var ${1:[type]} ${2:$value} {3}  */
      foreach ([] as $value) {}
      
      /** @var ${1:[type]} ${2:$value} {3}  */
      foreach ([] as $key => $value) {}
      ```
    - while
    ```php
     /** @var ${1:[type]} ${2:$value} {3}  */
     while ($value = array_pop($arrs)) {}
     ```
  - enum `Enum MyEnum{}` [PHP8.1-enumerations](https://wiki.php.net/rfc/enumerations)
  - const `const MyConst=1`

- getTypeFromValue
  - int
    - $var = 1_000_000 `@var int`
  - flaot
    - $var = 0.1_000_1 `@var float`
  - class
    - $var = new MyClass(); `@var MyClass`
    - $var = new class {}; `@var object`
  - closure
    - $var = function(){} `@var \Closure `
    - $var = fn()=>123 `@var \Closure `
  
- php
    - union types `Iterator|Countable` #150
    - constructor property promotion `__construct(private string $name)` #156
    - Trailing comma #153
    - Intersection types `Iterator&Countable` [PHP8.1-intersection_types](https://wiki.php.net/rfc/pure-intersection-types)
    - Readonly Properties `public readonly $uid` [PHP8.1-readonly_properties](https://wiki.php.net/rfc/readonly_properties_v2)
    - Param with default null #140

- other
  - Add `@mixin` tag #120
  - placeholder: `@param ${1:mixed} ${2:name} ${3:}` #106
  - replace tag (`@author`): Remove the cache and take effect in time after modifying the configuration
  - etc...

### CODE REVIEW

**Comment by tianyiw2013** [comment]
I'm optimizing the code. Please don't merge. ^_^

File: package.json [hunk 1]
```diff
@@ -39,13 +39,42 @@
     "commands": [
       {
         "command": "php-docblocker.trigger",
-        "title": "Insert PHP Docblock"
+        "title": "%Insert PHP Docblock%"
```
Comment (tianyiw2013) [inline package.json:42]: localization: package.nls.json

File: package.nls.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+    "Insert PHP Docblock": "Insert PHP Docblock",
+    "option.unknownType": "Placeholder of unknown type",
+    "option.withParameterDescription": "Use the parameter name as the description. e.g /** @param string $username username */",
+    "option.defaultMessage": "Document default message",
+    "option.defaultMessage.blank": "blank",
+    "option.defaultMessage.name": "function/class/var name",
+    "option.defaultMessage.undocumented": "e.g. Undocumented function",
+    "option.variableInline": "If enabled, generated doc in a line. e.g. /** @var mixed */",
+    "option.variableWithKey": "If enabled, generated document will contain a variable name. e.g. @var mixed $name"
+}
```
Comment (tianyiw2013) [inline package.nls.json:11]: My English is poor, please check and improve the content

File: src/completions.ts [hunk 3]
```diff
@@ -296,40 +289,37 @@ export default class Completions implements CompletionItemProvider
...
+     * Replace tag
+     * 
+     * @param {string} tag 
+     * @param {string} snippet 
+     * @returns {string}
      */
-    protected getTags():Array<{tag:string, snippet:string}>
+    protected replaceTagSnippet(tag:string, snippet:string): string
     {
-        if (!this.formatted) {
```
Comment (tianyiw2013) [inline src/completions.ts:321]: If it is marked as formatted, it cannot take effect immediately after the user modifies the author information

**Review by tianyiw2013 — COMMENTED** [review]
It's time to merge. Thank you

**Comment by shivapoudel** [comment]
@tianyiw2013 Thank you for this effort. @neild3r I was using her version of **PHPDocblocker2** because of the flexibility with this configuration, my friend :)

```
	"php-docblocker2.extra": [
		{
			"content": "@since ${###:1.0.0}",
			"after": "message",
			"gapAfter": true,
			"scope": "class,func",
		},
		{
			"content": "@since ${###:1.0.0}",
			"after": "message",
			"scope": "prop",
		},
	],
```

If this setting arrived at this extension then that will be much helpful. Both of you a warm thanks for rebasing all the stuffs 🤟

**Comment by tianyiw2013** [comment]
@shivapoudel, I will implement this feature in the my next pull request.

**Review by tianyiw2013 — COMMENTED** [review]
.

**Review by tianyiw2013 — COMMENTED** [review]
@neild3r Can be merged.

**Comment by shivapoudel** [comment]
@tianyiw2013 @neild3r Any updates regarding this https://github.com/neild3r/vscode-php-docblocker/pull/175#issuecomment-901347072

**Comment by neild3r** [comment]
@shivapoudel you can already do this without this being merged, the configuration is just slightly longer. For example for the function you can do

```json
    "php-docblocker.functionTemplate":     {
        "message": {},
        "since": {
            "content": "@since ${###:1.0.0}",
            "gapAfter": true
        },
        "param": {},
        "return": {},
        "extra": {}
    },
```

**Comment by shivapoudel** [comment]
@neild3r What about class and prop? Can you provide me the way out example?

**Comment by neild3r** [comment]
@shivapoudel 

```json
    "php-docblocker.propertyTemplate": {
        "message": {},
        "since": {
            "content": "@since ${###:1.0.0}",
            "gapAfter": true
        },
        "var": {},
        "extra": {}
    },
    "php-docblocker.classTemplate": {
        "message": {},
        "since": {
            "content": "@since ${###:1.0.0}",
            "gapAfter": true
        },
        "extra": {}
    },
```

**Comment by shivapoudel** [comment]
@neild3r Everything works great for implementing your above suggestions. By the way, is that simplified `php-docblocker2.extra` supported over this PR or not? Additionally, I would love to see `@param` description supported like this soon :)

image

**Comment by neild3r** [comment]
@shivapoudel that isn't in this PR and param descriptions is being worked on as part of #158

**Comment by shivapoudel** [comment]
@neild3r Ah! I found [this commit for param description](https://github.com/neild3r/vscode-php-docblocker/pull/161/commits/5b9d73a26b68c69c53c8a67a0cbb4dd7546c940a) in the PR #161. By the way, any plans to ship this feature in the upcoming release and when is it coming?

Thanks!

**Comment by neild3r** [comment]
Sorry yes I pointed to the wrong one. Probably within the next couple of weeks but I'm currently focusing on missing php syntax functionality which takes priority

**Comment by tianyiw2013** [comment]
I have withdrawn my published extension in the vscode market.
If you no longer need this PR, you can close it.
Thank you for your work. @neild3r

**Comment by neild3r** [comment]
@tianyiw2013 I will keep it open for reference. I intend on implementing many of the remaining parts of this PR. Thank you for all your hard work I appreciate you moving things forward and hope you will continue to contribute.

**Pull request closed without merging**